### ISSUE TITLE
performance: avoid file watchers

### ISSUE THREAD

**Issue by justinmk3** [aws/aws-toolkit-vscode#3216]
## Problem

Toolkit has various file watchers for SAM template.yaml files, CDK, etc. These can negatively affect performance.

## Solution

1. ~~Introduce settings to limit the cost of Toolkit file watchers:~~
1. ✅  exclude files/directories specified by vscode `search.exclude`, `files.exclude` settings. https://github.com/aws/aws-toolkit-vscode/pull/3770
1. ✅ Change `template.yaml` detect to be _on-demand_ instead of eager (omnipresent, filewatcher-driven)
    - When a `template.yaml` is needed but not found (such as "SAM run/debug", and "Sync SAM" command), mention this in the info message. (Related: [#3211](https://github.com/aws/aws-toolkit-vscode/pull/3211))
    - Implemented in #3931 + #3962

**Comment by justinmk3** [comment]
The "schema service" in particular can be a complete show-stopper if your workspace has many big projects (which... may contain... yaml files):

```
2023-03-27 19:46:16 [DEBUG]: schema service: handle yaml mapping: sam -> file:///projects/sam-installation/aws-sam-cli-src/tests/smoke/templates/sar/s3-uncompressor-template.yaml
2023-03-27 19:46:16 [DEBUG]: schema service: handle yaml mapping: sam -> file:///projects/sam-installation/aws-sam-cli-src/tests/smoke/templates/sar/alexa-skills-kit-color-expert-template.yaml
2023-03-27 19:46:16 [DEBUG]: schema service: handle yaml mapping: sam -> file:///projects/sam-installation/aws-sam-cli-src/tests/smoke/templates/sar/SecretsManagerRedshiftRotationSingleUser-template.yaml
2023-03-27 19:46:17 [DEBUG]: schema service: handle yaml mapping: sam -> file:///projects/sam-installation/aws-sam-cli-src/tests/smoke/templates/sar/magic-8-ball-template.yaml
2023-03-27 19:46:17 [DEBUG]: schema service: handle yaml mapping: sam -> file:///projects/sam-installation/aws-sam-cli-src/tests/smoke/templates/sar/SecretsManagerMongoDBRotationSingleUser-template.yaml
2023-03-27 19:46:17 [DEBUG]: schema service: handle yaml mapping: sam -> file:///projects/sam-installation/aws-sam-cli-src/tests/smoke/templates/sar/AmazonConnectSalesforceLambda-template.yaml
<100s more...>
```

- The above workspace has 40739 files and AWS Toolkit stalls 5+ minutes at startup (doesn't block the rest of the IDE but does block the Toolkit itself!).
- Easier to reproduce on a vm with a slow(ish) filesystem, such as a CodeCatalyst devenv.

**Comment by justinmk3** [comment]
Item 3 in the above list was implemented in #3931 + #3962.

Further performance work is tracked in:

- https://github.com/aws/aws-toolkit-vscode/labels/performance
- https://github.com/aws/aws-toolkit-vscode/issues/3510
